internal/jobservice/eventstojobs/manage_subs.go

```diff
@@ -0,0 +1,331 @@
+	}
+
+	_, err := jse.sqlJobService.UnsubscribeJobSet(jse.ctx, key.Queue, key.JobSetId)
+	return err
+}
+
+func (jse *JobSetSubscriptionExecutor) HasSubscription(key *repository.JobSetKey) bool {
+	jse.mutex.Lock()
+	defer jse.mutex.Unlock()
+
+	_, ok := jse.subscriptions[*key]
+	return ok
+}
+
+func (jse *JobSetSubscriptionExecutor) NumActiveSubscriptions() int {
+	jse.mutex.Lock()
+	defer jse.mutex.Unlock()
+
+	return len(jse.subscriptions)
+}
+
+func NewJobSetSubscription(ctx context.Context, eventReader events.JobEventReader, subInfo *repository.SubscribedTuple, subTimeout time.Duration, subDoneChan chan<- *repository.JobSetKey, sqlJobService repository.SQLJobService) *JobSetSubscription {
+	newCtx, cancel := context.WithCancel(ctx)
+	return &JobSetSubscription{
+		ctx:           newCtx,
+		cancel:        cancel,
+		eventReader:   eventReader,
+		JobSetKey:     subInfo.JobSetKey,
+		fromMessageId: subInfo.FromMessageId,
+		subTimeout:    subTimeout,
+		subDoneChan:   subDoneChan,
+		sqlJobService: sqlJobService,
+	}
+}
+
+func (js *JobSetSubscription) Subscribe() error {
+	requestFields := log.Fields{
+		"job_set_id":      js.JobSetId,
+		"queue":           js.Queue,
+		"from_message_id": js.fromMessageId,
+	}
+
+	defer func() {
+		js.subDoneChan <- &repository.JobSetKey{
+			Queue:    js.Queue,
+			JobSetId: js.JobSetId,
+		}
+	}()
+
+	log.WithFields(requestFields).Debugf("Calling GetJobEventMessage")
+
+	stream, err := js.eventReader.GetJobEventMessage(js.ctx, &api.JobSetRequest{
+		Id:            js.JobSetId,
+		Queue:         js.Queue,
+		Watch:         true,
+		FromMessageId: js.fromMessageId,
+	})
+	if err != nil {
+		log.WithFields(requestFields).WithError(err).Error("error from GetJobEventMessage")
+		js.cancel()
+		return err
+	}
+
+	log.WithFields(requestFields).Debug("Got stream")
+
+	g, _ := errgroup.WithContext(js.ctx)
+
+	// Subscription status check ticker.
+	timeout := time.NewTicker(js.subTimeout)
+
+	g.Go(func() error {
+		for {
+			select {
+			case <-js.ctx.Done():
+				return nil
+			case <-timeout.C:
+				log.WithFields(requestFields).Debug("JobSetSubscription.Subscribe checking subscription status")
+				// Stream is created with *our* context, therefore if we cancel, stream.Recv() should bail out too.
+				unsub, err := js.sqlJobService.CheckToUnSubscribe(js.ctx, js.Queue, js.JobSetId, js.subTimeout)
+				if err != nil {
+					log.WithFields(requestFields).WithError(err).Error("IsJobSetSubscribed error")
+				}
+				// We should unsubscribe
+				if unsub {
+					log.WithFields(requestFields).Info("subscription done")
+					js.cancel()
+					return nil
+				}
+			}
+		}
+	})
+
+	// Nanosecond, zero wait essentially.
+	g.Go(func() error {
+		nextRecv := time.After(1 * time.Nanosecond)
+
+		// this loop will run until the context is canceled
+		for {
+			select {
+			case <-js.ctx.Done():
+				log.WithFields(requestFields).Debug("context is done")
+				return nil
+			case <-nextRecv:
```

Comment:
Would changing `case <-nextRecv:` here to just a plain `default:` (and dropping the `nextRecv` var) work?

Reply:
It would work, but  if there's a recoverable stream error, we're going to hammer the `stream.Recv()` call until it stops erroring.

Reply:
I don't understand - with `nextRecv` being a 1 ns ticker, wouldn't that also do the same?

Reply:
No, its not a ticker, its a `time.After` which only fires once, and when it hits an error the loop sets the `nextRecv` to `time.After(5 * time.Seconds)`.

Reply:
Aha - yes, of course - my error. Now, ☕ .